internal/build/path_mapping.go

```diff
@@ -111,15 +113,40 @@ func fileToPathMapping(file string, mounts []model.Mount) (pathMapping, *PathMap
 		// need ospath.RealChild -- but then can't deal with deleted local files.
 		relPath, isChild := ospath.Child(m.LocalPath, file)
 		if isChild {
+			localPathIsFile, err := isFile(m.LocalPath)
+			if err != nil {
+				return pathMapping{}, pathMappingErrf("error stat'ing: %v", err)
+			}
+			var containerPath string
+			if endsWithSlash(m.ContainerPath) && localPathIsFile {
+				fileName := path.Base(m.LocalPath)
+				containerPath = filepath.Join(m.ContainerPath, fileName)
+			} else {
+				containerPath = filepath.Join(m.ContainerPath, relPath)
+			}
 			return pathMapping{
 				LocalPath:     file,
-				ContainerPath: filepath.Join(m.ContainerPath, relPath),
+				ContainerPath: containerPath,
 			}, nil
 		}
 	}
 	return pathMapping{}, pathMappingErrf("file %s matches no mounts", file)
 }
 
+// TODO(dmiller): this won't work in Windows
+func endsWithSlash(path string) bool {
+	return strings.HasSuffix(path, "/")
+}
+
+func isFile(path string) (bool, error) {
```

Comment:
This isn't quite what you want. First, it's buggy (this treats symlinks as non-files). I'm also not sure it's what you want semantically. What does docker do if you COPY dir1 dir2/? I thought it created dir2/dir1

Reply:
Good point about symlinks! Will fix.

I didn't change the handling of copying directories to directories. This should only take effect if the source is a file but the destination is a directory.

> What does docker do if you COPY dir1 dir2/? I thought it created dir2/dir1
Looks like it doesn't nest directories in that case.

```dockerfile
FROM iron/go:dev

RUN mkdir /remotedir2
COPY dir1 /remotedir2/
```

```
> tree
.
├── Dockerfile
├── dir1
│   └── hi
```

```
> docker run -it 303a3d0f2157 /bin/sh
/go # cd /tmp/
/tmp # ls
/tmp # cd /
/ # ls
bin         etc         home        media       proc        root        sbin        sys         usr
dev         go          lib         mnt         remotedir2  run         srv         tmp         var
/ # cd remotedir2/
/remotedir2 # ls
hi
```

Reply:
i would probably change this to be !isDir instead of isFile, which is probably closer to what we want?